File: internal/handler/commit/creates_commit.go
package commit

import (
	"fmt"
	"html/template"
	"net/http"
	"strconv"
	"strings"

	"github.com/gin-gonic/gin"
	"gitlab.com/gitlab-org/gitlab-redux/internal/model"
	"gitlab.com/gitlab-org/gitlab-redux/internal/service"
	"gitlab.com/gitlab-org/gitlab-redux/internal/util"
)

// CreatesCommit handles creating commits
type CreatesCommit struct {
	authService *service.AuthService
	userService *service.UserService
	repoService *service.RepositoryService
	mrService   *service.MergeRequestService
	logger      *util.Logger
}

// NewCreatesCommit creates a new instance of CreatesCommit
func NewCreatesCommit(
	authService *service.AuthService,
	userService *service.UserService,
	repoService *service.RepositoryService,
	mrService *service.MergeRequestService,
	logger *util.Logger,
) *CreatesCommit {
	return &CreatesCommit{
		authService: authService,
		userService: userService,
		repoService: repoService,
		mrService:   mrService,
		logger:      logger,
	}
}

// CreateCommit creates a commit
func (c *CreatesCommit) CreateCommit(
	ctx *gin.Context,
	service *service.CommitService,
	successPath string,
	failurePath interface{},
	failureView string,
	successNotice string,
	targetProject *model.Project,
) error {
	// Get the current user from the context
	currentUser, exists := ctx.Get("current_user")
	if !exists {
		return util.NewUnauthorizedError("user not authenticated")
	}
	user := currentUser.(*model.User)

	// Get the project from the context
	project, exists := ctx.Get("project")
	if !exists {
		return util.NewNotFoundError("project not found")
	}
	projectObj := project.(*model.Project)

	// Use the target project if provided, otherwise use the current project
	if targetProject == nil {
		targetProject = projectObj
	}

	// Get the branch name or ref
	branchNameOrRef := c.getBranchNameOrRef(ctx)

	// Check if the user can push to the branch
	userAccess := c.userService.GetUserAccess(ctx, user, targetProject)
	if userAccess.CanPushToBranch(branchNameOrRef) {
		// Set the project to commit into
		projectToCommitInto := targetProject
		differentProject := false
		branchName := branchNameOrRef
		ctx.Set("project_to_commit_into", projectToCommitInto)
		ctx.Set("different_project", differentProject)
		ctx.Set("branch_name", branchName)
	} else {
		// Get the user's fork of the target project
		fork := c.userService.GetForkOf(ctx, user, targetProject)
		if fork == nil {
			return util.NewForbiddenError("you don't have a fork of this project")
		}

		// Set the project to commit into
		projectToCommitInto := fork
		differentProject := true
		branchName := c.repoService.NextBranch(ctx, fork, "patch")
		ctx.Set("project_to_commit_into", projectToCommitInto)
		ctx.Set("different_project", differentProject)
		ctx.Set("branch_name", branchName)
	}

	// Get the start branch
	startBranch := c.getStartBranch(ctx)
	ctx.Set("start_branch", startBranch)

	// Get the commit parameters
	commitParams := c.getCommitParams(ctx)
	commitParams["start_project"] = ctx.Get("project_to_commit_into")
	commitParams["start_branch"] = startBranch
	commitParams["source_project"] = projectObj
	commitParams["target_project"] = targetProject
	commitParams["branch_name"] = ctx.Get("branch_name")

	// Execute the commit service
	result, err := service.New(ctx.Get("project_to_commit_into").(*model.Project), user, commitParams).Execute(ctx)
	if err != nil {
		return err
	}

	// Handle the result
	if result.Status == "success" {
		// Get the final success path
		finalSuccessPath := c.getFinalSuccessPath(ctx, successPath, targetProject)

		// Update the flash notice
		c.updateFlashNotice(ctx, successNotice, finalSuccessPath)

		// Handle the response format
		switch ctx.GetHeader("Accept") {
		case "application/json":
			ctx.JSON(http.StatusOK, gin.H{
				"message":   "success",
				"filePath": finalSuccessPath,
			})
		default:
			ctx.Redirect(http.StatusFound, finalSuccessPath)
		}
	} else {
		// Set the flash alert
		ctx.Set("flash_alert", c.formatFlashNotice(result.Message))

		// Get the failure path
		var finalFailurePath string
		switch p := failurePath.(type) {
		case string:
			finalFailurePath = p
		case func() string:
			finalFailurePath = p()
		default:
			finalFailurePath = "/"
		}

		// Handle the response format
		switch ctx.GetHeader("Accept") {
		case "application/json":
			ctx.JSON(http.StatusUnprocessableEntity, gin.H{
				"error":    result.Message,
				"filePath": finalFailurePath,
			})
		default:
			if failureView != "" {
				ctx.HTML(http.StatusUnprocessableEntity, failureView, nil)
			} else {
				ctx.Redirect(http.StatusFound, finalFailurePath)
			}
		}
	}

	return nil
}

// AuthorizeEditTree authorizes editing the tree
func (c *CreatesCommit) AuthorizeEditTree(ctx *gin.Context) error {
	// Get the project from the context
	project, exists := ctx.Get("project")
	if !exists {
		return util.NewNotFoundError("project not found")
	}
	projectObj := project.(*model.Project)

	// Get the branch name or ref
	branchNameOrRef := c.getBranchNameOrRef(ctx)

	// Check if the user can collaborate with the project
	if c.userService.CanCollaborateWithProject(ctx, projectObj, branchNameOrRef) {
		return nil
	}

	// Return an access denied error
	return util.NewForbiddenError("you don't have permission to edit this tree")
}

// formatFlashNotice formats a flash notice
func (c *CreatesCommit) formatFlashNotice(message string) string {
	// Replace newlines with <br> tags
	formattedMessage := strings.ReplaceAll(message, "\n", "<br>")

	// Sanitize the message
	return template.HTMLEscapeString(formattedMessage)
}

// updateFlashNotice updates the flash notice
func (c *CreatesCommit) updateFlashNotice(ctx *gin.Context, successNotice, successPath string) {
	// Create the changes link
	changesLink := fmt.Sprintf(`<a href="%s" class="gl-link">changes</a>`, successPath)

	// Create the default message
	defaultMessage := fmt.Sprintf("Your %s have been committed successfully.", changesLink)

	// Set the flash notice
	if successNotice != "" {
		ctx.Set("flash_notice", successNotice)
	} else {
		ctx.Set("flash_notice", defaultMessage)
	}

	// Check if we should create a merge request
	if c.shouldCreateMergeRequest(ctx) {
		// Check if the merge request exists
		if c.mergeRequestExists(ctx) {
			ctx.Set("flash_notice", nil)
		} else {
			// Get the different project flag
			differentProject, _ := ctx.Get("different_project")
			isDifferentProject := differentProject.(bool)

			// Create the merge request message
			var mrMessage string
			if isDifferentProject {
				mrMessage = "You can now submit a merge request to get this change into the original project."
			} else {
				mrMessage = "You can now submit a merge request to get this change into the original branch."
			}

			// Append the merge request message to the flash notice
			ctx.Set("flash_notice", ctx.Get("flash_notice").(string)+" "+mrMessage)
		}
	}
}

// getFinalSuccessPath gets the final success path
func (c *CreatesCommit) getFinalSuccessPath(ctx *gin.Context, successPath string, targetProject *model.Project) string {
	// Check if we should create a merge request
	if c.shouldCreateMergeRequest(ctx) {
		// Check if the merge request exists
		if c.mergeRequestExists(ctx) {
			return c.getExistingMergeRequestPath(ctx)
		} else {
			return c.getNewMergeRequestPath(ctx, targetProject)
		}
	}

	// Handle the success path
	switch p := successPath.(type) {
	case string:
		return p
	case func() string:
		return p()
	default:
		return "/"
	}
}

// getNewMergeRequestPath gets the path for a new merge request
func (c *CreatesCommit) getNewMergeRequestPath(ctx *gin.Context, targetProject *model.Project) string {
	// Get the project to commit into
	projectToCommitInto, _ := ctx.Get("project_to_commit_into")
	projectToCommitIntoObj := projectToCommitInto.(*model.Project)

	// Get the branch name
	branchName, _ := ctx.Get("branch_name")
	branchNameStr := branchName.(string)

	// Get the start branch
	startBranch, _ := ctx.Get("start_branch")
	startBranchStr := startBranch.(string)

	// Get the default merge request target
	defaultTarget := projectToCommitIntoObj.DefaultMergeRequestTarget

	// Create the merge request path
	return fmt.Sprintf(
		"/projects/%d/merge_requests/new?merge_request[target_project_id]=%d&merge_request[source_branch]=%s&merge_request[target_branch]=%s",
		projectToCommitIntoObj.ID,
		defaultTarget.ID,
		branchNameStr,
		startBranchStr,
	)
}

// getExistingMergeRequestPath gets the path for an existing merge request
func (c *CreatesCommit) getExistingMergeRequestPath(ctx *gin.Context) string {
	// Get the project from the context
	project, _ := ctx.Get("project")
	projectObj := project.(*model.Project)

	// Get the merge request from the context
	mergeRequest, _ := ctx.Get("merge_request")
	mergeRequestObj := mergeRequest.(*model.MergeRequest)

	// Create the merge request path
	return fmt.Sprintf("/projects/%d/merge_requests/%d", projectObj.ID, mergeRequestObj.ID)
}

// mergeRequestExists checks if a merge request exists
func (c *CreatesCommit) mergeRequestExists(ctx *gin.Context) bool {
	// Get the current user from the context
	currentUser, _ := ctx.Get("current_user")
	user := currentUser.(*model.User)

	// Get the project from the context
	project, _ := ctx.Get("project")
	projectObj := project.(*model.Project)

	// Get the project to commit into
	projectToCommitInto, _ := ctx.Get("project_to_commit_into")
	projectToCommitIntoObj := projectToCommitInto.(*model.Project)

	// Get the branch name
	branchName, _ := ctx.Get("branch_name")
	branchNameStr := branchName.(string)

	// Get the start branch
	startBranch, _ := ctx.Get("start_branch")
	startBranchStr := startBranch.(string)

	// Find the merge request
	mr := c.mrService.FindBySourceAndTarget(
		ctx,
		user,
		projectObj.ID,
		projectToCommitIntoObj.ID,
		branchNameStr,
		startBranchStr,
	)

	// Set the merge request in the context
	if mr != nil {
		ctx.Set("merge_request", mr)
	}

	return mr != nil
}

// shouldCreateMergeRequest checks if we should create a merge request
func (c *CreatesCommit) shouldCreateMergeRequest(ctx *gin.Context) bool {
	// Get the create merge request parameter
	createMR, exists := ctx.GetQuery("create_merge_request")
	if !exists {
		return false
	}

	// Parse the boolean value
	createMRBool, err := strconv.ParseBool(createMR)
	if err != nil {
		return false
	}

	// Get the different project flag
	differentProject, _ := ctx.Get("different_project")
	isDifferentProject := differentProject.(bool)

	// Get the start branch
	startBranch, _ := ctx.Get("start_branch")
	startBranchStr := startBranch.(string)

	// Get the branch name
	branchName, _ := ctx.Get("branch_name")
	branchNameStr := branchName.(string)

	// Check if we should create a merge request
	return createMRBool && (isDifferentProject || startBranchStr != branchNameStr)
}

// getBranchNameOrRef gets the branch name or ref
func (c *CreatesCommit) getBranchNameOrRef(ctx *gin.Context) string {
	// Get the branch name
	branchName, exists := ctx.Get("branch_name")
	if exists {
		return branchName.(string)
	}

	// Get the ref
	ref, exists := ctx.Get("ref")
	if exists {
		return ref.(string)
	}

	return ""
}

// getStartBranch gets the start branch
func (c *CreatesCommit) getStartBranch(ctx *gin.Context) string {
	// Get the ref
	ref, exists := ctx.Get("ref")
	if exists {
		return ref.(string)
	}

	// Get the branch name
	branchName, exists := ctx.Get("branch_name")
	if exists {
		return branchName.(string)
	}

	return ""
}

// getCommitParams gets the commit parameters
func (c *CreatesCommit) getCommitParams(ctx *gin.Context) map[string]interface{} {
	// Get the commit parameters from the context
	commitParams, exists := ctx.Get("commit_params")
	if exists {
		return commitParams.(map[string]interface{})
	}

	return make(map[string]interface{})
}

File: rust_app/src/controllers/concerns/creates_commit.rs
use actix_web::{web, HttpResponse};
use std::sync::Arc;

use crate::models::project::Project;
use crate::models::user::User;
use crate::services::commit_service::CommitService;
use crate::utils::access_control::AccessControl;
use crate::utils::flash::Flash;
use crate::utils::merge_request::MergeRequestFinder;
use crate::utils::security::SecurityUtils;

/// Module for handling commit creation
pub trait CreatesCommit {
    /// Create a commit using the specified service
    fn create_commit(
        &self,
        service: Box<dyn CommitService>,
        success_path: String,
        failure_path: String,
        failure_view: Option<String>,
        success_notice: Option<String>,
        target_project: Option<Arc<Project>>,
    ) -> HttpResponse {
        let target_project = target_project.unwrap_or_else(|| self.project().clone());
        let branch_name_or_ref = self.branch_name_or_ref();

        let (project_to_commit_into, different_project, branch_name) = if self
            .user_access(&target_project)
            .can_push_to_branch(&branch_name_or_ref)
        {
            (target_project.clone(), false, branch_name_or_ref.clone())
        } else {
            let fork = self.current_user().fork_of(&target_project);
            let next_branch = fork.repository().next_branch("patch");
            (fork, true, next_branch)
        };

        let start_branch = self.ref_name().unwrap_or_else(|| branch_name.clone());

        let mut commit_params = self.commit_params().clone();
        commit_params.insert(
            "start_project".to_string(),
            project_to_commit_into.id().to_string(),
        );
        commit_params.insert("start_branch".to_string(), start_branch);
        commit_params.insert(
            "source_project".to_string(),
            self.project().id().to_string(),
        );
        commit_params.insert(
            "target_project".to_string(),
            target_project.id().to_string(),
        );
        commit_params.insert("branch_name".to_string(), branch_name);

        let result = service.execute(&project_to_commit_into, self.current_user(), commit_params);

        match result.status() {
            "success" => {
                let success_path = self.final_success_path(&success_path, &target_project);
                let success_notice = self.update_flash_notice(success_notice, &success_path);

                self.flash().set_notice(success_notice);

                HttpResponse::Found()
                    .header("Location", success_path)
                    .finish()
            }
            _ => {
                let error_message = self.format_flash_notice(result.message());
                self.flash().set_alert(error_message);

                let failure_path = if failure_path.contains("{{") {
                    // Handle dynamic paths with template variables
                    self.evaluate_path_template(&failure_path)
                } else {
                    failure_path
                };

                if let Some(view) = failure_view {
                    // Render the specified view
                    self.render_view(&view)
                } else {
                    // Redirect to failure path
                    HttpResponse::Found()
                        .header("Location", failure_path)
                        .finish()
                }
            }
        }
    }

    /// Authorize edit tree operation
    fn authorize_edit_tree(&self) -> Result<(), actix_web::Error> {
        if self.can_collaborate_with_project(self.project(), &self.branch_name_or_ref()) {
            Ok(())
        } else {
            Err(actix_web::error::ErrorForbidden("Access denied"))
        }
    }

    /// Format flash notice message
    fn format_flash_notice(&self, message: &str) -> String {
        let formatted_message = message.replace("\n", "<br>");
        self.sanitize_html(&formatted_message, &["br"])
    }

    /// Update flash notice with success message
    fn update_flash_notice(&self, success_notice: Option<String>, success_path: &str) -> String {
        let changes_link = format!("<a href=\"{}\" class=\"gl-link\">changes</a>", success_path);

        let default_message = format!("Your {} have been committed successfully.", changes_link);

        let mut notice = success_notice.unwrap_or(default_message);

        if self.create_merge_request() {
            if !self.merge_request_exists() {
                let mr_message = if self.different_project() {
                    "You can now submit a merge request to get this change into the original project."
                } else {
                    "You can now submit a merge request to get this change into the original branch."
                };

                notice.push_str(&format!(" {}", mr_message));
            }
        }

        notice
    }

    /// Get the final success path
    fn final_success_path(&self, success_path: &str, target_project: &Arc<Project>) -> String {
        if self.create_merge_request() {
            if self.merge_request_exists() {
                self.existing_merge_request_path()
            } else {
                self.new_merge_request_path(target_project)
            }
        } else {
            success_path.to_string()
        }
    }

    /// Get path for new merge request
    fn new_merge_request_path(&self, target_project: &Arc<Project>) -> String {
        format!("/projects/{}/merge_requests/new", target_project.id())
    }

    /// Get path for existing merge request
    fn existing_merge_request_path(&self) -> String {
        format!(
            "/projects/{}/merge_requests/{}",
            self.project().id(),
            self.merge_request().id()
        )
    }

    /// Check if merge request exists
    fn merge_request_exists(&self) -> bool {
        let finder = MergeRequestFinder::new(self.current_user());
        let params = vec![("project_id", self.project().id().to_string())];
        !finder.execute(params).is_empty()
    }

    /// Check if merge request should be created
    fn create_merge_request(&self) -> bool {
        // Even if the field is set, if we're checking the same branch
        // as the target branch in the same project,
        // we don't want to create a merge request.
        let create_mr = self
            .params()
            .get("create_merge_request")
            .and_then(|v| v.parse::<bool>().ok())
            .unwrap_or(false);

        create_mr && (self.different_project() || self.branch_name_or_ref() != self.target_branch())
    }

    /// Get branch name or ref
    fn branch_name_or_ref(&self) -> String {
        self.branch_name()
            .unwrap_or_else(|| self.ref_name().unwrap_or_default())
    }

    // Required trait methods that need to be implemented by the controller
    fn project(&self) -> Arc<Project>;
    fn current_user(&self) -> Arc<User>;
    fn user_access(&self, project: &Arc<Project>) -> Box<dyn AccessControl>;
    fn ref_name(&self) -> Option<String>;
    fn branch_name(&self) -> Option<String>;
    fn target_branch(&self) -> String;
    fn commit_params(&self) -> std::collections::HashMap<String, String>;
    fn flash(&self) -> &dyn Flash;
    fn params(&self) -> std::collections::HashMap<String, String>;
    fn different_project(&self) -> bool;
    fn merge_request(&self) -> Arc<crate::models::merge_request::MergeRequest>;
    fn can_collaborate_with_project(&self, project: &Arc<Project>, ref_name: &str) -> bool;
    fn sanitize_html(&self, html: &str, allowed_tags: &[&str]) -> String;
    fn render_view(&self, view: &str) -> HttpResponse;
    fn evaluate_path_template(&self, template: &str) -> String;
}
